src/context/mod.rs

```diff
@@ -119,6 +126,16 @@ impl<'a> SourceTextProvider<'a> for FileRunContext<'a, '_> {
     }
 }
 
+impl<'a> NodeParentProvider<'a> for FileRunContext<'a, '_> {
+    fn maybe_node_parent(&self, node: Node<'a>) -> Option<Node<'a>> {
+        self.node_parent_cache.get(&node).copied()
```

Comment:
Currently the cache is "fully populated" so we can just do a simple lookup and return what we find

But if wanted to optimize further an idea would be to just cache parents of "named nodes" and then I guess here fall back to `Node::parent()`?